### PR TITLE
Make NodeAnnouncement creation conditional on startup

### PR DESCRIPTION
**Pull request by blico**
PR for [#274](https://github.com/lightningnetwork/lnd/issues/274)

With this change, on startup we craft a new `NodeAnnoucement` based on the relevant `cfg` options and compare that with the one on disk. The node announcement in the database is only updated if the new node announcement differs from the old one. 

Testing: 

- Added a test covering node announcement on startup functionality to `graph_topology_notifications` within `lnd_test`.

- Added a unit test for `IsEqual` within `node_announcement_test`

### CODE REVIEW

File: lnd_test.go [hunk 1]
```diff
@@ -2744,7 +2744,7 @@ func testGraphTopologyNotifications(net *networkHarness, t *harnessTest) {
 
 	// We should receive an update advertising the newly connected node,
 	// Bob's new node announcement, and the channel between Bob and Carol.
-	for i := 0; i < 3; i++ {
+	for i := 0; i < 4; i++ {
```
Comment (blico) [inline lnd_test.go]: I noticed that we are receiving two channel announcements in the previous test. This change gets my test working, but I am not sure if it is hiding something that should be a failure.

File: lnwire/node_announcement_test.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	
+	// Ensure CompareNodes returns false when the nodes' 
+	// public keys are not equal.
+	*node2 = *node1
+	priv2, _ := btcec.NewPrivateKey(btcec.S256())
+	node2.NodeID = priv2.PubKey()
+	if node1.CompareNodes(node2) {
+		t.Fatalf("expected node comparison to return false")
+	}
+}
```
Comment (halseth) [inline lnwire/node_announcement_test.go]: newline

File: server.go [hunk 3]
```diff
@@ -203,47 +203,79 @@ func newServer(listenAddrs []string, chanDB *channeldb.DB, cc *chainControl,
...
-	selfNode.AuthSig, err = discovery.SignAnnouncement(s.nodeSigner,
-		s.identityPriv.PubKey(), nodeAnn,
-	)
-	if err != nil {
-		return nil, fmt.Errorf("unable to generate signature for "+
-			"self node announcement: %v", err)
+
+	updateNodeAnn := true
+
+	chanGraph := chanDB.ChannelGraph()
```
Comment (halseth) [inline server.go]: I would move these two declarations to right before where they are used the first time

File: lnd_test.go (hunk 1 above)
Comment (halseth) [reply lnd_test.go]: What channels are you getting announcements for?

File: lnd_test.go [hunk 4]
```diff
@@ -2790,6 +2790,58 @@ func testGraphTopologyNotifications(net *networkHarness, t *harnessTest) {
...
+
+	// First restart Bob with no new config changes
+	if err := net.RestartNode(net.Bob, nil); err != nil {
+		t.Fatalf("unable to restart Bob's node: %v", err)
+	}
+
+	if err := net.ConnectNodes(ctxb, net.Alice, net.Bob); err != nil {
+		t.Fatalf("unable to connect alice to bob: %v", err)
+	}
+
```
Comment (halseth) [inline lnd_test.go]: I would add a check here to ensure we did not get any new announcements.

File: lnd_test.go [hunk 5]
```diff
@@ -2790,6 +2790,58 @@ func testGraphTopologyNotifications(net *networkHarness, t *harnessTest) {
...
+					numBobAnnRecv++
+					if numBobAnnRecv > 1 {
+						t.Fatalf("unexpected amount of node updates received from Bob")
+					}
+				default:
+					t.Fatalf("unknown node update pubey: %v",
+						nodeUpdate.IdentityKey)
+				}
+			}
+		case <-time.After(time.Second * 10):
```
Comment (halseth) [inline lnd_test.go]: Will this block for 10 seconds also in the success case? I think it is better to check that we get the expected announcement then do a quick wait (~1 sec?) to make sure we are not getting any more.

File: lnwire/node_announcement.go [hunk 6]
```diff
@@ -152,3 +152,43 @@ func (a *NodeAnnouncement) DataToSign() ([]byte, error) {
 
 	return w.Bytes(), nil
 }
+
+// CompareNodes compares the configurable fields within two NodeAnnouncement
+// objects, and returns whether they are equal or not.
+func (a *NodeAnnouncement) CompareNodes(b *NodeAnnouncement) bool {
```
Comment (halseth) [inline lnwire/node_announcement.go]: Maybe call it `IsEqual`?

File: lnd_test.go (hunk 1 above)
Comment (blico) [reply lnd_test.go]: I am no longer getting an extra announcement, so I changed the number back to 3.

File: lnd_test.go (hunk 5 above)
Comment (Roasbeef) [reply lnd_test.go]: @halseth not that we require longer timeouts when running on Travis due to the jitter and slowness of the machines.

File: networktest.go [hunk 7]
```diff
@@ -185,6 +185,7 @@ func (l *lightningNode) genArgs() []string {
 	args = append(args, fmt.Sprintf("--configfile=%v", l.cfg.DataDir))
 	args = append(args, fmt.Sprintf("--adminmacaroonpath=%v", l.cfg.AdminMacPath))
 	args = append(args, fmt.Sprintf("--readonlymacaroonpath=%v", l.cfg.ReadMacPath))
+	args = append(args, fmt.Sprintf("--externalip=%v", l.cfg.ExternalIPs))
```
Comment (Roasbeef) [inline networktest.go]: This isn't necessary, as the `ExternalIPs` parameter is never set (in normal test operation). Also, assigning the `--externalip` to an empty string as done here, may trigger some odd behavior in the tests due to the way resolution is currently attempted. As an example, here's the result of me setting `--externalip` twice locally (once with `localhost` and the other with no parameter):
```
{
    "nodes": [
        {
            "last_update": 1506398714,
            "pub_key": "02a196c6106ba977647119414dd8da019a708cdeed2b8ff7e1c691b0aa329643ba",
            "alias": "02a196c6106ba9776471\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000",
            "addresses": [
                {
                    "network": "tcp",
                    "addr": "127.0.0.1:9735"
                },
                {
                    "network": "tcp",
                    "addr": "[107:1:59c9:d1fa::]:9735"
                }
            ]
        }
    ],
    "edges": [
    ]
}

```

I see you added this in order to be able to exercise this new behavior within the integration tests. But I'm not convinced it's really necessary as this is such a trivial feature and the added tests will slow down the execution of all the integration tests a bit (until we move to parallelizing all test execution).

Comment (blico) [reply networktest.go]: I agree that the test is superfluous. My latest commit reverts the integration test. Let me know if any other changes are needed.

File: server.go [hunk 8]
```diff
@@ -203,47 +203,77 @@ func newServer(listenAddrs []string, chanDB *channeldb.DB, cc *chainControl,
...
-		PubKey:               privKey.PubKey(),
-		Alias:                alias.String(),
-		Features:             globalFeatures,
-	}
 
-	// If our information has changed since our last boot, then we'll
-	// re-sign our node announcement so a fresh authenticated version of it
-	// can be propagated throughout the network upon startup.
-	//
-	// TODO(roasbeef): don't always set timestamp above to _now.
```
Comment (jimpo) [inline server.go]: Was the TODO dropped intentionally?

File: server.go [hunk 9]
```diff
@@ -203,47 +203,77 @@ func newServer(listenAddrs []string, chanDB *channeldb.DB, cc *chainControl,
...
 	}
 
-	chanGraph := chanDB.ChannelGraph()
-
 	// TODO(roasbeef): make alias configurable
 	alias, err := lnwire.NewNodeAlias(hex.EncodeToString(serializedPubKey[:10]))
 	if err != nil {
 		return nil, err
 	}
-	selfNode := &channeldb.LightningNode{
```
Comment (jimpo) [inline server.go]: I kind of like having this up here. It makes more sense to me that a NodeAnnouncement would be derived from a LightningNode than constructing it directly. In fact, `LightningNode` could have a method that constructs a `NodeAnnouncement`, which you could use to get both `nodeAnn` and `oldNodeAnn`.

File: lnwire/node_announcement.go [hunk 10]
```diff
@@ -152,3 +152,43 @@ func (a *NodeAnnouncement) DataToSign() ([]byte, error) {
 
 	return w.Bytes(), nil
 }
+
+// IsEqual compares the configurable fields within two NodeAnnouncement
```
Comment (jimpo) [inline lnwire/node_announcement.go]: Would add here that the Signature is not compared.

File: server.go [hunk 11]
```diff
@@ -203,47 +203,77 @@ func newServer(listenAddrs []string, chanDB *channeldb.DB, cc *chainControl,
...
+
+		if err := chanGraph.SetSourceNode(selfNode); err != nil {
+			return nil, fmt.Errorf("can't set self node: %v", err)
+		}
+
+		nodeAnn.Signature = selfNode.AuthSig
 	}
 
-	nodeAnn.Signature = selfNode.AuthSig
 	s.currentNodeAnn = nodeAnn
```
Comment (jimpo) [inline server.go]: This now does not have a valid signature. Is that a problem?

**Comment by Roasbeef** [comment]
Closing in favor of #1120 where all the remaining review comments will be addressed, and original source commits retained.

**Pull request closed without merging**